src/goto-instrument/cover.cpp

```diff
@@ -218,6 +218,299 @@ void collect_operands(const exprt &src, std::vector<exprt> &dest)
+
+Function: replacement_and_conjunction
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: To replace the i-th expr of ''operands'' with each
+          expr inside ''replacement_exprs''.
+
+\*******************************************************************/
+
+std::set<exprt> replacement_and_conjunction(
+  const std::set<exprt> &replacement_exprs,
+  const std::vector<exprt> &operands,
+  const std::size_t i)
+{
+  std::set<exprt> result;
+  for(auto &y : replacement_exprs)
+  {
+    std::vector<exprt> others;
+    for(std::size_t j=0; j<operands.size(); j++)
+      if(i!=j)
+        others.push_back(operands[j]);
+
+    others.push_back(y);
+    exprt c=conjunction(others);
+    result.insert(c);
+  }
+  return result;
+}
+        
+/*******************************************************************\
+
+Function: non_ordered_expr_expansion
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: To expand all non ordered predicates within an expr
+
+\*******************************************************************/
+
+std::set<exprt> non_ordered_expr_expansion(const exprt &src)
+{
+  std::set<exprt> result;
+
+  std::set<exprt> s1, s2;
+
+  if(src.id()!=ID_not||!is_arithmetic_predicate(src.op0()))
+    s1.insert(src);
+  else
+  {
+    auto res=ordered_negation(src.op0());
+    s1.insert(res.begin(), res.end());
+  }
+
+  // expand negations and non-ordered predicates
+  bool changed=true;
+  while(changed)
+  {
+    changed=false;
+    for(auto &x: s1)
+    {
+      std::vector<exprt> operands;
+      collect_operands(x, operands);
+      for(std::size_t i=0; i<operands.size(); i++)
+      {
+        std::set<exprt> res;
+        if(operands[i].id()==ID_not)
+        {
+          exprt no=operands[i].op0();
+          if(is_arithmetic_predicate(no))
+          {
+            res=ordered_negation(no);
+            if(!res.empty()) changed=true;
+          }
+        }
+        else
+        {
+          if(operands[i].id()==ID_le
+             || operands[i].id()==ID_ge
+             || operands[i].id()==ID_notequal)
+          {
+            res=non_ordered_predicate_expansion(operands[i]);
+            if(!res.empty()) changed=true;
+          }
+        }
+        std::set<exprt> co=replacement_and_conjunction(res, operands, i);
+        s2.insert(co.begin(), co.end());
+        if(!res.empty()) break;
+      }
+      if(!changed) s2.insert(x);
+    }
+    if(!changed) break;
+    s1.swap(s2);
+    s2.clear();
+  }
+
+  return s1;
+}
+
+/*******************************************************************\
+
+Function: decision_expansion
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+std::set<exprt> decision_expansion(const exprt &dec)
+{
+  std::set<exprt> ctrl;
+  // dec itself may be a non-ordered predicate
+  exprt d(dec);
+  if(d.id()==ID_not) d=d.op0();
+  if(is_arithmetic_predicate(d))
+  {
+    auto res=non_ordered_predicate_expansion(d);
```

Comment:
How about non_ordered_predicate_expansion just returns its argument when there was nothing to expand? That would avoid those subsequent if(empty) handling.